lib/firenest/replicated_server.ex

```diff
@@ -0,0 +1,306 @@
+
+  alias Firenest.Topology
+
+  @type state() :: term()
+
+  @callback init(arg :: term()) ::
+              {:ok, state}
+              | {:ok, state(), timeout() | :hibernate}
+              | {:stop, reason :: term()}
+
+  @callback replica_up(Topology.node_ref(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @callback replica_down(Topology.node_ref(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @callback handle_call(request :: term(), GenServer.from(), state()) ::
+              {:reply, reply, new_state}
+              | {:reply, reply, new_state, timeout() | :hibernate}
+              | {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason, reply, new_state}
+              | {:stop, reason, new_state}
+            when reply: term(), new_state: state(), reason: term()
+
+  @callback handle_cast(request :: term(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @callback handle_info(request :: term(), state()) ::
+            {:noreply, new_state}
+            | {:noreply, new_state, timeout() | :hibernate}
+            | {:stop, reason :: term(), new_state}
+          when new_state: state()
+
+  @callback handle_replicated(request :: term(), Topology.node_ref(), state()) ::
+              {:noreply, new_state}
+              | {:noreply, new_state, timeout() | :hibernate}
+              | {:stop, reason :: term(), new_state}
+            when new_state: state()
+
+  @callback terminate(reason, state()) :: term()
+            when reason: :normal | :shutdown | {:shutdown, term()}
+
+  @callback format_status(reason, pdict_and_state :: list()) :: term()
+            when reason: :normal | :terminate
+
+  @callback code_change(old_vsn, state :: term(), extra :: term()) ::
+              {:ok, new_state :: term()} | {:error, reason :: term()} | {:down, term()}
+            when old_vsn: term()
+
+  @optional_callbacks [terminate: 2, code_change: 3, format_status: 2]
+
+  def start_link(mod, arg, opts \\ []) do
+    name = Keyword.fetch!(opts, :name)
+    topology = Keyword.fetch!(opts, :topology)
+    GenServer.start_link(__MODULE__, {mod, arg, topology, name}, opts)
+  end
+
+  def call(server, msg, timeout \\ 5_000) do
+    GenServer.call(server, msg, timeout)
+  end
+
+  def cast(server, msg) do
+    GenServer.cast(server, msg)
+  end
+
+  def reply(from, reply) do
+    GenServer.reply(from, reply)
+  end
+
+  def send(to_node_ref, msg) do
+    {topology, name, node_ref} = Process.get(__MODULE__)
```

Comment:
Is the process dictionary taking place of the ets dispatch table used previously? It's not clear yet why it's needed in this case

Reply:
This is used only within the same process to make it more convenient to do sends and broadcasts. It is not tracking any sort of state between processes.

Reply:
It's just for convenience. This will be called only from inside the process from the callback module when we want to send/broadcast to another node. Unfortunately, inside the callback module we don't know all the data that is needed to send the message. I saw three solutions:
* pass some kind of "token" to callbacks that would contain this information (similar to phoenix channels and the socket struct)
* introduce additional return types from callbacks to perform the actions (e.g. `{:send, msg, state}`)
* save this in pdict and retrieve when necessary.

I went with option 3 since I believe this leads to the simplest code. It might appear a bit "magical" especially when calling `SyncedServer.broadcast(msg)` with no destination, but I think this is fine.